modules/apps/product-navigation/product-navigation-control-menu-api/src/main/java/com/liferay/product/navigation/control/menu/BaseJSPProductNavigationControlMenuEntry.java

```diff
@@ -72,9 +72,7 @@ public boolean includeIcon(
 			httpServletRequest, httpServletResponse, getIconJspPath());
 	}
 
-	public void setServletContext(ServletContext servletContext) {
-		_servletContext = servletContext;
-	}
+	protected abstract ServletContext getServletContext();
```

Comment:
@W2jx If the `ServletContext` is only used in the base class to obtain the `RequestDispatcher`, I think you can just let the subclass implement an abstract method `getRequestDispatcher(String jspPath)`. So subclasses still reference the `ServletContext` but directly give the `RequestDispatcher` to parent.

Reply:
cc @tinatian

Reply:
I will fix it.